### PR TITLE
Re-factor simulation instance creation

### LINKED ISSUE #299: Allow fluid simulation class to be embedded in other apps
One of the main goals of this refactoring is to allow the simulation to use any given command buffer, device and resolution.

- Implement a static `Create` function to initialize an instance of the simulator with error handling for resource creation.
- Use an app-provided command buffer instead of having a fixed one.
- Allow access to intermediate textures.

### PR DESCRIPTION
**Pull request by dnovillo**
This is the first step to fix #299.  It moves the creation of the simulation instance into a static method and starts disentangling it from the application.

Further cleanups will remove the ppx::Application link and the command buffer creation.

### CODE REVIEW

File: projects/fluid_simulation/shaders.h [hunk 1]
```diff
@@ -89,7 +89,7 @@ class Texture
     }
 
 private:
-    ProjApp* GetApp() const;
+    ppx::Application* GetApp() const;
```
Comment (chaoticbob) [inline projects/fluid_simulation/shaders.h:92]: Is it necessary for the FLuidSim to know about app?

Also, you might be aware already, there's static method `Application::Get()` that returns the current app instance. You can use this to get both the Application and Device.

File: projects/fluid_simulation/sim.cpp [hunk 2]
```diff
@@ -28,32 +28,32 @@ const ppx::grfx::Format kR    = ppx::grfx::FORMAT_R16_FLOAT;
...
+ppx::Result FluidSimulation::Initialize()
 {
     // Create descriptor pool shared by all pipelines.
     ppx::grfx::DescriptorPoolCreateInfo dpci = {};
     dpci.sampler                             = 1024;
     dpci.sampledImage                        = 1024;
     dpci.uniformBuffer                       = 1024;
     dpci.storageImage                        = 1024;
-    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateDescriptorPool(&dpci, &mDescriptorPool));
+    PPX_CHECKED_CALL(GetDevice()->CreateDescriptorPool(&dpci, &mDescriptorPool));
```
Comment (chaoticbob) [inline projects/fluid_simulation/sim.cpp:45]: This seems like it can be leaner since we know the number of simulation frames and the number of resources we'll use.

File: projects/fluid_simulation/sim.cpp [hunk 3]
```diff
@@ -238,7 +238,7 @@ void FluidSimulation::InitBloomTextures()
     ppx::int2 res = GetResolution(GetConfig().pBloomResolution->GetValue());
     mBloomTexture = std::make_unique<Texture>(this, "bloom", res.x, res.y, kRGBA);
     PPX_ASSERT_MSG(mBloomTextures.empty(), "Bloom textures already initialized");
-    for (uint32_t i = 0; i < GetConfig().pBloomIterations->GetValue(); i++) {
+    for (int i = 0; i < GetConfig().pBloomIterations->GetValue(); i++) {
```
Comment (chaoticbob) [inline projects/fluid_simulation/sim.cpp:241]: Curious why the change from `uint32_t` to `int`?

File: projects/fluid_simulation/sim.h [hunk 4]
```diff
@@ -67,13 +67,34 @@ class ProjApp;
 class FluidSimulation
 {
 public:
-    FluidSimulation(ProjApp* app);
-    ProjApp*                     GetApp() const { return mApp; }
-    const SimulationConfig&      GetConfig() const;
+    FluidSimulation(ppx::Application* app, ppx::grfx::DevicePtr device, ppx::uint2 resolution, const SimulationConfig& config)
```
Comment (chaoticbob) [inline projects/fluid_simulation/sim.h:70]: IHMO: I don't think it's necessary to store a pointer to the application. The device I can understand.

File: projects/fluid_simulation/shaders.h (hunk 1 above)
Comment (dnovillo) [reply projects/fluid_simulation/shaders.h:92]: No it is not, but that requires more refactoring. I wanted to make it in incremental steps to facilitate reviews.

File: projects/fluid_simulation/sim.cpp (hunk 3 above)
Comment (dnovillo) [reply projects/fluid_simulation/sim.cpp:241]: Warning about different signedness. The knob returns int.

File: projects/fluid_simulation/sim.h (hunk 4 above)
Comment (dnovillo) [reply projects/fluid_simulation/sim.h:70]: Yeah. I'll be removing application ties next.

File: projects/fluid_simulation/sim.cpp (hunk 2 above)
Comment (dnovillo) [reply projects/fluid_simulation/sim.cpp:45]: Yup. You want this change in this PR or a subsequent one? This is the first of several cleanups.

**Review by chaoticbob — APPROVED** [review]
